From 2efb8d98d620480cd4cea252cbeb1bdf3670bdc8 Mon Sep 17 00:00:00 2001
From: Calvin Johnson <calvin.johnson@nxp.com>
Date: Mon, 30 Apr 2018 11:40:01 +0530
Subject: [PATCH 330/741] staging: fsl_ppfe/eth: unregister netdev after
 pfe_phy_exit

commit 8f7ac9d9ff39eebae98ab797820f644408740a6e from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

rmmod pfe.ko throws below warning:

kernfs: can not remove 'phydev', no directory
------------[ cut here ]------------
WARNING: CPU: 0 PID: 2230 at fs/kernfs/dir.c:1481
kernfs_remove_by_name_ns+0x90/0xa0

This is caused when the unregistered netdev structure is accessed to
disconnect phy.

Resolve the issue by unregistering netdev after disconnecting phy.

Signed-off-by: Calvin Johnson <calvin.johnson@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/staging/fsl_ppfe/pfe_eth.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/staging/fsl_ppfe/pfe_eth.c b/drivers/staging/fsl_ppfe/pfe_eth.c
index c4f8348c6f1c..6af21ac7bbcb 100644
--- a/drivers/staging/fsl_ppfe/pfe_eth.c
+++ b/drivers/staging/fsl_ppfe/pfe_eth.c
@@ -2464,15 +2464,15 @@ static void pfe_eth_exit_one(struct pfe_eth_priv_s *priv)
 {
 	netif_info(priv, probe, priv->ndev, "%s\n", __func__);
 
-	if (!us) {
+	if (!us)
 		pfe_eth_sysfs_exit(priv->ndev);
 
-		unregister_netdev(priv->ndev);
-	}
-
 	if (!(priv->einfo->phy_flags & GEMAC_NO_PHY))
 		pfe_phy_exit(priv->ndev);
 
+	if (!us)
+		unregister_netdev(priv->ndev);
+
 	if (priv->mii_bus)
 		pfe_eth_mdio_exit(priv->mii_bus);
 
-- 
2.17.1

